File: src/emulator/io.nim
import bitops
import utils
import types

var
    # Serial Data Buffer
    SB: uint8
    SC: uint8

    # Interrupt Registers
    IF*: uint8
    IE*: uint8

    # Timer Registers
    DIV*: uint16
    TIMA*: uint8
    TMA*: uint8
    TAC*: uint8

    # PPU Registers
    LCDC*: uint8
    STAT*: uint8
    SCY*: uint8
    SCX*: uint8
    LY*: uint8
    LYC*: uint8
    DMA*: uint8
    BGP*: uint8
    OBP0*: uint8
    OBP1*: uint8
    WY*: uint8
    WX*: uint8

    # PPU Internal Registers
    LX*: uint8
    WLY*: uint8

    booting*: bool = true

var
    serialRegTable: array[2, ptr uint8] = [addr SB, addr SC]
    timerRegTable: array[3, ptr uint8] = [addr TIMA, addr TMA, addr TAC]
    ppuRegTable: array[12, ptr uint8] = [addr LCDC, addr STAT, addr SCY,
            addr SCX, addr LY, addr LYC, addr DMA, addr BGP, addr OBP0,
            addr OBP1,
            addr WY, addr WX]

proc getIoReg*(address: int): uint8 =
    if address.isboundto(0x01, 0x02):
        result = serialRegTable[address - 1][]

    elif address == 0x0F: result = IF
    elif address == 0x04: result = msb(DIV)

    elif address.isboundto(0x05, 0x07):
        result = timerRegTable[address - 5][]

    elif address.isboundto(0x40, 0x4B):
        result = ppuRegTable[address and 0xF][]

proc setIoReg*(address: int, data: uint8): void =
    if address.isboundto(0x01, 0x02):
        serialRegTable[address - 1][] = data

    elif address == 0x0F: IF = data
    elif address == 0x04: DIV = 0

    elif address.isboundto(0x05, 0x07):
        timerRegTable[address - 5][] = data

    elif address.isboundto(0x40, 0x4B):
        ppuRegTable[address and 0xF][] = data

    elif address == 0x50:
        booting = false

proc getLCDC*(lct: LCDCType): bool = return LCDC.testBit(lct.ord)
proc getLCDS*(lst: LCDSType): bool = return STAT.testBit(lst.ord)

proc sendIntReq*(ISR: IntType): void =
    setBit(IF, ISR.ord)

proc `LCDS=`*(lst: LCDSType, value: bool): void = 
    if value: STAT.setBit(lst.ord) 
    else: STAT.clearBit(lst.ord)

proc setMode*(mode: PPUStateType): void =
    STAT.clearBits(0, 1)
    STAT = STAT or mode.ord.uint8

proc getTileMapBase*(win: bool = false): uint16 =
    let flag = if win: getLCDC(WINTILEMAPAREA) else: getLCDC(BGTILEMAPAREA)
    return if flag: 0x9C00 else: 0x9800

proc getTileDataBase*(): uint16 = 
    return if getLCDC(BGANDWINTILEDATAAREA): 0x8000 else: 0x9000

File: src/io.c
#include "io.h"

#include <stdbool.h>
#include <stddef.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>

#include "joypad.h"
#include "logger.h"
#include "types.h"
#include "utils.h"

IORegisters ioRegs = {0};
#define IOREGBASE ((uint8_t *)&ioRegs)
#define IOREGLEN sizeof(IORegisters)

static inline uint8_t getRegisterAt(uint8_t address, long pageStart) {
    uint8_t *regAddr = (IOREGBASE + address + (uintptr_t)pageStart);
    if ((regAddr >= IOREGBASE) && (regAddr < (IOREGBASE + IOREGLEN))) {
        return *regAddr;
    } else {
        logState(FATAL,
                 "Attempted to READ from an invalid IO Register address.");
        return 0xFF;
    }
}

static inline void setRegisterAt(uint8_t address, long pageStart,
                                 uint8_t data) {
    uint8_t *regAddr = (IOREGBASE + address + (uintptr_t)pageStart);
    if ((regAddr >= IOREGBASE) && (regAddr < (IOREGBASE + IOREGLEN))) {
        *regAddr = data;
    } else {
        logState(FATAL,
                 "Attempted to WRITE to an invalid IO Register address.");
    }
}

uint8_t getIoReg(uint8_t address) {
    switch (address) {
        case 0x00:
            setKeyMask();
            if (js.fallingEdge) {
                sendIntReq(INTJOYPAD);
            }
            return js.keyMask;

        case 0x04:
            return MSB(ioRegs.DIV);

        case 0x0F:
            return ioRegs.IF;

        default:
            if (address >= 0x01 && address <= 0x07) {
                return getRegisterAt(address, offsetof(IORegisters, PAGE_0));
            } else if (address >= 0x40 && address <= 0x4B) {
                return getRegisterAt((address - 0x40) + 1,
                                     offsetof(IORegisters, PAGE_1));
            } else {
                return 0xFF;
            }
            break;
    }
}

void setIoReg(uint8_t address, uint8_t data) {
    switch (address) {
        case 0x00:
            js.P1 = data;
            break;

        case 0x04:
            ioRegs.DIV = 0;
            break;

        case 0x0F:
            ioRegs.IF = data;
            break;

        case 0x50:
            ioRegs.booting = false;
            break;

        default:
            if (address >= 0x01 && address <= 0x07) {
                setRegisterAt(address, offsetof(IORegisters, PAGE_0), data);
            } else if (address >= 0x40 && address <= 0x4B) {
                setRegisterAt((address - 0x40) + 1,
                              offsetof(IORegisters, PAGE_1), data);
            }
    }
}

bool getLCDC(LCDCType lct) { return BT(ioRegs.LCDC, lct); }

bool getLCDS(LCDSType lst) { return BT(ioRegs.STAT, lst); }

uint16_t getTileMapBase(bool win) {
    bool flag = win ? getLCDC(WINTILEMAPAREA) : getLCDC(BGTILEMAPAREA);
    return flag ? 0x9C00 : 0x9800;
}

uint16_t getTileDataBase(void) {
    return getLCDC(BGANDWINTILEDATAAREA) ? 0x8000 : 0x9000;
}
